### ISSUE TITLE
[New Concept Exercise]:  Functions in Python

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3121]
This issue describes how to implement  the  `functions` **concept exercise** for the Python track.
There is an outdated issue for this exercise with some discussion.  It can be found [here](https://github.com/exercism/python/issues/2353)
The related **concept documents** can be found [here](https://github.com/exercism/python/tree/main/concepts/functions).

<br>

## ✅ Getting started

_If you have not yet created or contributed to a concept exercise, this issue will require some upfront reading to give you the needed background knowledge. Some good example exercises to look at in the repo:_

<details>
<summary>💡<b>Example Exercises</b>💡 (<em>click to expand</em>)</summary>
<br>

   1. [Little Sister's Vocabulary](https://github.com/exercism/python/tree/main/exercises/concept/little-sisters-vocab)
   2. [Meltdown Mitigation](https://github.com/exercism/python/tree/main/exercises/concept/meltdown-mitigation) 
   3. [Making the Grade](https://github.com/exercism/python/tree/main/exercises/concept/making-the-grade)
   4. [Ellen's Alien Game](https://github.com/exercism/python/tree/main/exercises/concept/ellens-alien-game)

</details>

We also recommend completing one or more of the concept exercises (_they're called "learning exercises"_) [on the website](https://exercism.org/tracks/python/concepts). 

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please go through the following documents:

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concept Exercises**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concept Exercises?](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Stories](https://exercism.org/docs/building/tracks/stories)

## 🎯 Goal

This `functions` concept exercise is meant to teach a deeper understanding and use of `functions`  in Python. It should also explain how Python treats/views functions (_as callable objects_), and dig into some of the features that make Python `functions` unique.

<br>

## 💡Learning objectives

- Understand more about  Python `scopes` and `namespaces`
  - understand the difference between the `global` and `nonlocal` keywords and when to use them
- Get familiar with the [special attributes](https://docs.python.org/3/reference/datamodel.html#types) of Python functions
- Get familiar with best practices when using `return`, and the difference between _explicit_ and _implicit_ `return`
   - functions without an explicit return will return the singleton object `None`
- Understand what is meant by _"functions are first class **objects** in Python"_.
  - understand that `functions` **are** `objects` in Python, and that they have `types`
  - understand that `functions` can be assigned to variables, used in expressions,  and stored in various data structures such as `dicts` or `lists`
  - create `functons` that are assigned to variables, used in expressions, and stored in different data structures.
  - understand and create` functions` that are/can be nested inside one another
- Understand that Python considers a function a form of  `callable object`. 
- Understand that a user-defined `function object` is created by a `function definition`.

<br>

## 🤔 Concepts

- `callable objects`
- `first-class functions`
- `global`
- `nested functions`
- `nonlocal`
- `return`, `implicit return`, `explicit return`
- `scope`
-  special `function attributes`

<br>

## 🚫 Topics that are Out of scope

<br>
<details>
<summary><b>Concepts & Subjects that are Out of Scope</b> (<em>click to expand</em>)</summary>
<br>

- `named parameters` (_these can be touched on if needed_)
-  `default parameters` (_these can be touched on, if needed_)
- `arbitrary parameters`
- `*args & **kwargs`
- `keyword-only arguments`
- `/` and `*` for requiring parameter types
-  `functions-as-arguments` (_this can be mentioned, but shouldn't be required for the exercise_)
- `functions-as-returns`(_this can be mentioned, but will be covered in-depth in `higher-order functions`)
- `closures` (_these will be covered in a different exercise_)
- `decorators` (_these will be covered in a different exercise_)
-  `functools.wraps` (_this is used mostly for decorators_)
-  `functools` (_this will get its own exercise_)
- `comprehensions`
- `generators`
- `lambda`, `anonymous functions` (_these will be covered in a different exercise_)
- `recursion`

</details>
<br>

## ↩️  Prerequisites

_These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  Since `functions` is a "meta" topic,  these will probably need to be adjusted to fit the parameters of the exercise._ 

<details>
<summary>Prereqs  (<em>click to expand</em>)</summary>
<br>

- `basics`
- `bools`
- `comparisons`
- `lists`
- `list-methods`
- `loops`
- `numbers`
- `strings`
- `string-methods`

</details>

<br>

##  📚 Resources for Writing and Reference

<details>
<summary><b>Resources</b> (<em>click to expand</em>)</summary>
<br>

- [Python Docs: Naming and Binding](https://docs.python.org/3/reference/executionmodel.html#naming-and-binding)
- [Python Docs: Python Scopes and Namespaces](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces)
- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Function definitions (Python Library Reference)](https://docs.python.org/3/reference/compound_stmts.html#function-definitions)
- [Dan Bader:  Python s Functions are First-Class](https://dbader.org/blog/python-first-class-functions)
- [Real Python:  Python Scope & the LEGB Rule](https://realpython.com/python-scope-legb-rule/)
- [Stack Overflow: Short Description of Python's Scoping Rules?](https://stackoverflow.com/questions/291978/short-description-of-the-scoping-rules)
- [Real Python: Python Inner Functions: What are they Good For?](https://realpython.com/inner-functions-what-are-they-good-for/)
- [Stack Abuse: How to Use global and nonlocal variables in Python](https://stackabuse.com/how-to-use-global-and-nonlocal-variables-in-python/)
- [The `nonlocal` Statement (Python Docs)](https://docs.python.org/3/reference/simple_stmts.html#nonlocal)
- [The `global` Statement (Python Docs)](https://docs.python.org/3/reference/simple_stmts.html#global)
- [Real Python: The Python return Statement: Usage and Best Practices](https://stackabuse.com/how-to-use-global-and-nonlocal-variables-in-python/)
- [Calls (Python Library Reference)](https://docs.python.org/3/reference/expressions.html#calls)
- [Is it a Class or a Function?  It's a Callable! (Trey Hunner)](https://treyhunner.com/2019/04/is-it-a-class-or-a-function-its-a-callable/)
- [Callables in Python (Vimal A.R Blog)](https://arvimal.blog/2017/08/09/callables-in-python/)
- [Functions as Objects in Python](https://medium.com/python-pandemonium/function-as-objects-in-python-d5215e6d1b0d) -- _this is a subscription-based medium blog, so not good for hints or links_
- [Python Datamodel: Types](https://docs.python.org/3/reference/datamodel.html#types)

</details>
 
<br>

### Exercise Ideas & Stories

Should you need inspiration for an exercise story, you can find a collection [here](https://exercism.org/docs/building/tracks/stories).  You can also port an exercise from another track, **but please make sure to only to include tasks that actually make sense in Python** and that add value for a student.  Remove/replace/add tasks as needed to make the concept clear/workable.

<br>

##  📁 Exercise Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b> (<em>click to collapse</em>)</summary>
<table>
<td>
<br>

* ### Exercise `introduction.md`

  For more information, see [**Exercise** `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsintroductionmd)

  - This can summarize/paraphrase the linked concept documents if they have already been created (either the `about` or the `introduction`).   The summary does need to have enough information and examples  for the student to complete all the tasks outlined for this concept exercise.

* ### Exercise `instructions.md`

   For more information, see [`instructions.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsinstructionsmd)

   Instructions for an exercise usually center on a story that sets up the code challenge to be solved.  You can create your own story, or fork one from the ones listed [here](https://exercism.org/docs/building/tracks/stories).  Please make sure to give credit to the original authors if you use a story or fork an exercise.

* ### Exercise `Exemplar.py` Solution
  For more information, see [exemplar implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-exemplar-implementation).
  
  This file should not use syntax or datas structures not introduced in this exercise or in this exercise's prerequisites.  It will be used as an "ideal" solution for the challenge, so make sure it conforms to PEP8 and other formatting conventions, and **does not use single letter variable names**.  It should also include proper module and function-level docstrings.  However, it should **NOT** include typehinting or type aliases.

* ### `<Exercise>.py` (Stub) for Implementation
  For more information, see [stub implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-stub-implementation).
  
  This file should provide the expected function names imported for testing, and optionally TODO comments and or docstrings to aid the student in their implementation.  TODOs and docstrings are not required.

* ### `<Exercise>_Test.py` Files
  For more information, see [Tests](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-tests).
  Additionally, please note that Python associates exercise tasks to tests via a [Pytest Marker](https://docs.pytest.org/en/7.1.x/example/markers.html), and uses [`unittest subtests`](https://docs.python.org/3/library/unittest.html#subtests) as a form of test paramaterization.  See the test file for [`Little Sisters Vocab`](https://github.com/exercism/python/blob/main/exercises/concept/little-sisters-vocab/strings_test.py) for examples of how these techniques work.

* ### Exercise `Hints.md`

  For more information on writing hints see [`hints.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docshintsmd)

  - Hints should provide enough information to get most students "un-stuck" and moving toward a solution.  They should **not** provide a student with a direct solution.
  - You can refer to one or more of the resources linked in this issue above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### Exercise Metadata Files Under  `.meta/config.json`

  For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#metadata-files)

   * `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
   * `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

<br>
</td>
</table>
</details>
<br>

## ♾️ Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [`config.json`](https://exercism.org/docs/building/tracks/config-json).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** :  `9f38f70d-48f4-4934-ae6e-d8277ede8edc`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue    

<br>

## 🎶 Implementation Notes

- As a reminder, code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**  We run all our `examplar.py` files through PyLint, but do not strictly require module docstrings.  We **do** require function docstrings similar to [PEP257](https://www.python.org/dev/peps/pep-0257/).  See [this concept exercise `exemplar.py`](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/.meta/exemplar.py) for an example.
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered either in the introduction to this exercise, or to one of its prerequisites.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.  

    - All asserts should contain a "user friendly" failure message (_these will display on the webiste to students, so be as clear as you can_). 
    - We use a `PyTest custom mark` to link test cases to exercise task numbers.  
    - We also use `unittest.subtest` to parameterize test input where/when needed.
Here is an [example testfile](https://github.com/exercism/python/blob/main/exercises/concept/making-the-grade/loops_test.py) that shows all three of these in action.

- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest-specific test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.

**Comment by BethanyG** [comment]
**Comments for discussion ported from original issue:**

@BethanyG  Any idea for what you want to see for the exercise here? Any analogous exercise you might want transcribed from another track? I can start looking at the tracks I'm familiar with. Nothing comes to mind right now.

_Originally posted by @bobahop in https://github.com/exercism/python/issues/2353#issuecomment-1157775990_

-----------

Rust uses lasagna, but we use lasagna for basics. Go uses Booking Up For Beauty, but that also makes significant use of string formatting, so it might not be as focused for functions as desired.

_Originally posted by @bobahop in https://github.com/exercism/python/issues/2353#issuecomment-1157788607_

------------

@bobahop there are a few things to note here:

1.   This issue has very outdated links and descriptions and it also pings the attached users every time we make a comment, so I think its best if we close it in favor of a new issue using the new template.  I will do that shortly.   

2. I have a strong preference for original stories/exercises, unless the port/concept is an almost 1-to-1, as was the case with `bools`.  That **isn't** to say that porting is verboten, but I'd much rather have  a purpose-built exercise than a shoehorned one.  We have had trouble in the past with contributors trying to port something and then getting stalled because they were not able to transfer some core piece to Python.

3.   As I've mentioned, this exercise is quite far "down" the exercise tree. It should not be considered an exercise on how to _make_ functions in Python, but rather as a setup to both `higher order functions` and related concepts.  So we should try to think of something that deepens a student's understanding/use of functions, and exercises some of the features unique to Python.  Otherwise, this may as well be a practice exercise or a concept-only topic.

I'll add some thoughts/ideas to the new issue.

_Originally posted by @BethanyG in https://github.com/exercism/python/issues/2353#issuecomment-1157841911_

**Comment by BethanyG** [comment]
So this wouldn't be a port of the tasks, but if you like the `lasagna` story, [`lasagna master`](https://github.com/exercism/go/tree/main/exercises/concept/lasagna-master/.docs) from **go** could be used and the tasks replaced/adapted/added to.

[Secrets](https://github.com/exercism/elixir/blob/main/exercises/concept/secrets/.docs/instructions.md) on the **Elixir** track looks like it has a good setup, and that the tasks might be adaptable to Python scenarios.  They will probably need a lot of replacing, but could work.

Likewise, [Library Fees](https://github.com/exercism/elixir/blob/main/exercises/concept/library-fees/.docs/instructions.md) could be adapted, but the instructions and introduction to the exercise would have to be amended to cover whatever `datetime` functions or operations were needed for completion.

[New Passport](https://github.com/exercism/elixir/blob/main/exercises/concept/new-passport/.docs/instructions.md) could be adapted, but not to use `with`, because that really really belongs in its own exercise.  However, I think the tasks could be changed to make it about assigning functions to variables, calling functions from other functions, nesting, and scope.  It could also be renamed "adventures at the DMV" and made ... dystopian ... where the story is about getting in and getting out of the DMV in the fastest way possible, or avoiding cranky employees by finding and excuse to be in the "quick processing" line.

[Pizza Pricing](https://github.com/exercism/fsharp/tree/main/exercises/concept/pizza-pricing/.docs) holds some promise, but again - the tasks would have to be re-done for Python.

Really, you could use just about any story [here](https://exercism.org/docs/building/tracks/stories).  The problematic bit is coming up with 4-6 tasks that cover the concept. 

If sudden purpose-built inspiration strikes, I'll update.

**Comment by bobahop** [comment]
Sound like many promising leads. I'm a bit remiss on researching the tracks, as I'm also slowly slogging through crafting interpreters. Very slowly.

**Comment by bobahop** [comment]
I like the dystopian DMV idea... a lot.

**Comment by BethanyG** [comment]
Creating an "Adventures at the DMV" story would have the added benefit of carrying a story credit (_and a legendary badge_).  😉

**Inspiration**:  https://www.youtube.com/watch?v=ONFj7AYgbko.  You could even borrow (with credit) the premise from Zootopia - the DMV is populated by Sloths, and any mistake not only sends you to a new line...but then you have to deal with the Sloths. 

Might need some brief background on what DMV stands for, but I think globally most people understand horrible bureaucracy, so it wouldn't take much.

**Comment by bobahop** [comment]
I think I'll start to work up some files tomorrow. And you're getting author credit, too, since it's your story idea.

What do you think of the title "Della's Delays at the DMV"?

**Comment by BethanyG** [comment]
"Della's Delays at the DMV"  is _**great**_.  A personal ambition of mine is to have as many alliterative concept exercise titles as possible.

**Comment by bobahop** [comment]
I just looked over the learning objectives and concepts and I winced a bit at `global` and `nonlocal`. Is it okay to include them in a way that shows how they can be problematical? Or should we leave them just in the about.md and not give the student experience in them as if they are something encouraged to be done?

**Comment by BethanyG** [comment]
Apologies.  Had some things to take care of this afternoon, so am late with replying to things.

> _I just looked over the learning objectives and concepts and I winced a bit at `global` and `nonlocal.` Is it okay to include them in a way that shows how they can be problematical?_ 

I think for the _exercise_, it's important to provide students the opportunity to practice using both `global` and `nonlocal` in an appropriate or "safe" fashion (_heavy guardrails_).  It doesn't have to be elaborate, and it doesn't have to be much more than a basic mimic (_I am thinking of something similar to how we set up [error handling](https://github.com/exercism/python/issues/2563) messaging in a lot of practice exercises_ ). 

And I think for the instructions/introduction, it would be great to put cautions in, and examples of why it can be a bad idea.  Although my thinking here is that if the task or tasks set up a "proper" or "good" use case, the students will more than likely find the foot-gun, and that will convey a really good lesson, about scope, using the keywords, AND about how the Python ethos is to say _"we are all adults here, so you need to think through your choices, Python's not going to do that for you."_   

It does seem somewhat "unfair" to set up students like that ... but I don't know that we teach them fluency by hiding the warts of the language.  While we do do that to some extent early on, by the time students are getting ready to feed functions to other functions, use recursion, nest functions, and all the other goodness that comes along with FP paradigms in Python, they should have a working sense of caution, and be building a sense of why endless mutability can be problematic.

So my thought is that maybe there is a task that has them reach out and modify/update a `global` constant.  Or _create_ a `global` in a function that other functions could read/access (and modify at peril).  I'm less clear on the `nonlocal`, although that could be done with an inner function -- and would be a quite legitimate (if minority) use-case.

...and thinking out loud here..... I don't know if we want to consider a `mutability` topic, or if we want to (_like we have with error handling_) insert messages and reminders about the dangers of mutability as we build out the exercises in the `functions` and `higher order functions` cluster ... as well as the `classes`, `class-customization`, `class-interfaces` cluster.

**Comment by BethanyG** [comment]
While stale, this issue should remain open.

**Comment by BethanyG** [comment]
Arg.  This keeps falling off the radar.  Will do better this week!

**Comment by bobahop** [comment]
I still think it is a good idea, but I am just focused on other things. I also have an idea for a practice exercise, but that is on the back burner as well.

**Comment by BethanyG** [comment]
I won't have time before this weekend I don't think - but I am going to PR turning off the stale bot, since none of us on the track are moving very fast these days (which is OK!).  Right now, if we comment on anything and 22 days pass, the critter wakes up and flags things as abandoned.  It's starting to not be useful, so.  😀

**Comment by BethanyG** [comment]
And no worries about doing other stuff - I just didn't want the bot to auto-close something that someone is still working on.
